clients/venice-producer/src/main/java/com/linkedin/venice/producer/AbstractVeniceProducer.java

```diff
@@ -0,0 +1,428 @@
+      VersionCreationResponse versionCreationResponse,
+      Properties veniceWriterProperties) {
+    int amplificationFactor = versionCreationResponse.getAmplificationFactor();
+    Integer partitionCount = versionCreationResponse.getPartitions();
+    Properties partitionerProperties = new Properties();
+    partitionerProperties.putAll(versionCreationResponse.getPartitionerParams());
+    VenicePartitioner venicePartitioner = PartitionUtils.getVenicePartitioner(
+        versionCreationResponse.getPartitionerClass(),
+        amplificationFactor,
+        new VeniceProperties(partitionerProperties));
+    return constructVeniceWriter(
+        veniceWriterProperties,
+        new VeniceWriterOptions.Builder(versionCreationResponse.getKafkaTopic()).setPartitioner(venicePartitioner)
+            .setPartitionCount(partitionCount)
+            .setChunkingEnabled(false)
+            .build());
+  }
+
+  // Visible for testing
+  protected VeniceWriter<byte[], byte[], byte[]> constructVeniceWriter(
+      Properties properties,
+      VeniceWriterOptions writerOptions) {
+    return new VeniceWriterFactory(properties).createVeniceWriter(writerOptions);
+  }
+
+  protected RecordSerializer<Object> getSerializer(Schema schema) {
+    return FastSerializerDeserializerFactory.getAvroGenericSerializer(schema);
+  }
+
+  private static Schema getSchemaFromObject(Object object) {
+    if (object instanceof GenericContainer) {
+      GenericContainer obj = (GenericContainer) object;
+      return obj.getSchema();
+    } else if (object instanceof CharSequence) { // convenience option.
+      return STRING_SCHEMA;
+    } else if (object instanceof Integer) {
+      return INT_SCHEMA;
+    } else if (object instanceof Long) {
+      return LONG_SCHEMA;
+    } else if (object instanceof Double) {
+      return DOUBLE_SCHEMA;
+    } else if (object instanceof Float) {
+      return FLOAT_SCHEMA;
+    } else if (object instanceof byte[] || object instanceof ByteBuffer) {
+      return BYTES_SCHEMA;
+    } else if (object instanceof Boolean) {
+      return BOOL_SCHEMA;
+    } else {
+      throw new VeniceException(
+          "Venice Producer only supports Avro objects, and primitives, found object of class: "
+              + object.getClass().toString());
+    }
+  }
+
+  @Override
+  public CompletableFuture<DurableWrite> asyncPut(K key, V value) {
+    return asyncPutInternal(APP_DEFAULT_LOGICAL_TS, key, value);
+  }
+
+  @Override
+  public CompletableFuture<DurableWrite> asyncPut(long logicalTime, K key, V value) {
+    if (logicalTime < 0) {
+      return getFutureCompletedExceptionally("Logical time must be a non-negative value. Got: " + logicalTime);
+    }
+
+    return asyncPutInternal(logicalTime, key, value);
+  }
+
+  private CompletableFuture<DurableWrite> asyncPutInternal(long logicalTime, K key, V value) {
+    String error = validateProducer();
+    if (!StringUtils.isEmpty(error)) {
+      return getFutureCompletedExceptionally(error);
```

Comment:
Does it make sense to record metrics around this?

Reply:
I don't think so. This should occur when the producer starts or is being shutdown (probably only during development). Users would also log the error on their end too when they encounter failed futures